### PR TITLE
feat(spanner): add support for batchWrite

### PR DESCRIPTION
**Pull request by alkatrivedi**
This PR contains Batch Write API implementation in Spanner.

### CODE REVIEW

File: src/transaction.ts [hunk 1]
```diff
@@ -2493,6 +2467,128 @@ promisifyAll(Transaction, {
...
+ *    .on('error', console.error)
+ *    .on('data', response => {
+ *          console.log('response: ', response);
+ *      })
+ *     .on('end', () => {
+ *          console.log('Request completed successfully');
+ *      });
+ * ```
+ */
+
```
Comment (surbhigarg92) [inline src/transaction.ts:2558]: nit: Remove this empty line

File: src/database.ts [hunk 2]
```diff
@@ -3210,6 +3212,110 @@ class Database extends common.GrpcServiceObject {
...
+   *  on 'data' event.
+   *
+   * @example
+   * ```
+   * const {Spanner} = require('@google-cloud/spanner');
+   * const spanner = new Spanner();
+   *
+   * const instance = spanner.instance('my-instance');
+   * const database = instance.database('my-database');
+   * const mutationGroups = new MutationGroup();
```
Comment (surbhigarg92) [inline src/database.ts:3235]: nit: `mutationGroups` to `mutationGroup`

File: system-test/spanner.ts [hunk 3]
```diff
@@ -4876,6 +4877,25 @@ describe('Spanner', () => {
         queryStreamMode(done, PG_DATABASE, query, POSTGRESQL_EXPECTED_ROW);
       });
 
+      // test for BATCH WRITE
```
Comment (surbhigarg92) [inline system-test/spanner.ts:4880]: Is this feature not available for PGSQL ? If available add a test for PG/

Comment (surbhigarg92) [inline system-test/spanner.ts:4880]: nit: No need for the comment. Test name should be self explanatory . Change the test name to

`GOOGLE_STANDARD_SQL should execute mutation group using Batch write'

File: src/database.ts [hunk 4]
```diff
@@ -3515,6 +3621,7 @@ class Database extends common.GrpcServiceObject {
 promisifyAll(Database, {
   exclude: [
     'batchTransaction',
+    'batchWrite',
```
Comment (surbhigarg92) [inline src/database.ts:3624]: Why do we need to add this method here, this method does not return a promise ?

File: src/database.ts [hunk 5]
```diff
@@ -3536,6 +3643,7 @@ callbackifyAll(Database, {
     'create',
     'batchCreateSessions',
     'batchTransaction',
+    'batchWrite',
```
Comment (surbhigarg92) [inline src/database.ts:3646]: Same Why do we need to add this method here, this method does not takes a callback ?

File: test/database.ts [hunk 6]
```diff
@@ -50,6 +50,7 @@ const fakePfy = extend({}, pfy, {
     promisified = true;
     assert.deepStrictEqual(options.exclude, [
       'batchTransaction',
+      'batchWrite',
```
Comment (surbhigarg92) [inline test/database.ts:53]: Unit tests should be added for batchwrite

File: system-test/spanner.ts (hunk 3 above)
Comment (alkatrivedi) [reply system-test/spanner.ts]: Done

File: src/database.ts (hunk 4 above)
Comment (alkatrivedi) [reply src/database.ts:3624]: No, this return a readable stream. Hence, adding it here

File: src/database.ts (hunk 5 above)
Comment (alkatrivedi) [reply src/database.ts:3646]: Yeah, it doesn' takes a callback directly

File: test/database.ts (hunk 6 above)
Comment (alkatrivedi) [reply test/database.ts:53]: Done

File: test/database.ts [hunk 7]
```diff
@@ -573,6 +574,94 @@ describe('Database', () => {
...
+    beforeEach(() => {
+      fakePool = database.pool_;
+      fakeSession = new FakeSession();
+      fakeDataStream = through.obj();
+
+      getSessionStub = (
+        sandbox.stub(fakePool, 'getSession') as sinon.SinonStub
+      ).callsFake(callback => callback(null, fakeSession));
+
+      requestStreamStub = sandbox
```
Comment (surbhigarg92) [inline test/database.ts:608]: Where are we using this `requestStreamStub`

File: test/database.ts [hunk 8]
```diff
@@ -573,6 +574,94 @@ describe('Database', () => {
     });
   });
 
+  describe('batchWrite', () => {
```
Comment (surbhigarg92) [inline test/database.ts:577]: Add a test case for "mutationGroups" passed as empty list or null.

File: src/database.ts (hunk 4 above)
Comment (surbhigarg92) [reply src/database.ts:3624]: Got it. Its an exclude list.

File: src/database.ts (hunk 5 above)
Comment (surbhigarg92) [reply src/database.ts:3646]: Got it. Its an exclude list.

File: test/database.ts (hunk 8 above)
Comment (alkatrivedi) [reply test/database.ts:578]: Done

File: test/database.ts (hunk 7 above)
Comment (alkatrivedi) [reply test/database.ts:614]: Added a test for it

**Review by surbhigarg92 — APPROVED** [review]

File: test/database.ts [hunk 9]
```diff
@@ -573,6 +576,138 @@ describe('Database', () => {
...
+        code: grpc.status.NOT_FOUND,
+        message: 'Session not found',
+      } as grpc.ServiceError;
+      let retryCount = 0;
+
+      database
+        .batchWrite(mutationGroups, options)
+        .on('data', () => {})
+        .on('error', err => {
+          console.log('err: ', err);
```
Comment (olavloite) [inline test/database.ts:688]: remove?

File: system-test/spanner.ts [hunk 10]
```diff
@@ -4876,6 +4877,42 @@ describe('Spanner', () => {
         queryStreamMode(done, PG_DATABASE, query, POSTGRESQL_EXPECTED_ROW);
       });
 
+      it('GOOGLE_STANDARD_SQL should execute mutation group using Batch write', function (done) {
```
Comment (olavloite) [inline system-test/spanner.ts:4880]: We don't seem to have any tests that cover the error handling of `batchWrite` (other than `Session not found`). We shoud add that? Either as an end-to-end test, or as a test with the Spanner mock server?

It would be good to have verification that:
1. The error handling works as expected if for example 1 mutation group fails, but the other X mutation groups in the request succeed. But also if there are multiple errors and multiple successes? Or only failures?
2. The events that are emitted by the stream that we return to the user are logical. E.g. does it only emit one `error` event and then stop? Or multiple `error` events if there are multiple errors? And what happens to the `done` event in those cases?

**Review by olavloite — COMMENTED** [review]
This PR does not appear to add any tests for error handling. Error handling of `BatchWrite` is probably one of the more difficult parts of this feature, as a request can return both successes and failures, and we should verify that our API handles that in a logical way.

File: src/transaction.ts [hunk 11]
```diff
@@ -2493,6 +2467,127 @@ promisifyAll(Transaction, {
...
+/**
+ * Formats the mutations.
+ *
+ * @param {string} method - CRUD method (insert, update, etc.).
+ * @param {string} table - Table to perform mutations in.
+ * @param {object | object[]} keyVals - Hash of key-value pairs representing the rows.
+ * @returns {spannerClient.spanner.v1.Mutation} - The formatted mutation.
+ * @throws {GoogleError} - If a row does not contain the correct number of columns.
+ */
+function buildMutation(
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply src/transaction.ts:2479]: This method is not part of any class. Private methods can only be created within a class scope. Also, this method is begin accessed from two classes "Transaction" and "MutationGroup" class. Hence, we can not add this method as private method to any specific class.

This method is part of module scope and we are not exporting this method.

File: test/database.ts (hunk 9 above)
Comment (alkatrivedi) [reply test/database.ts]: Done

File: src/transaction.ts [hunk 12]
```diff
@@ -2493,6 +2467,127 @@ promisifyAll(Transaction, {
...
+  };
+  const mutation: spannerClient.spanner.v1.IMutation = {
+    delete: {table, keySet},
+  };
+  return mutation as spannerClient.spanner.v1.Mutation;
+}
+
+/**
+ * This class represents a group of mutations that provides guaranteed atomicity and
+ * consistency across several CRUD operations.
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply src/transaction.ts]: Done

File: src/transaction.ts [hunk 13]
```diff
@@ -2493,6 +2467,127 @@ promisifyAll(Transaction, {
...
+}
+
+/**
+ * Formats a delete mutation.
+ *
+ * @param {string} table - The name of the table.
+ * @param {Key[]} keys - The keys for the rows to delete.
+ * @returns {spannerClient.spanner.v1.Mutation} - The formatted delete mutation.
+ */
+function buildDeleteMutation(
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply src/transaction.ts:2517]: This method is not part of any class. Private methods can only be created within a class scope. Also, this method is begin accessed from two classes "Transaction" and "MutationGroup" class. Hence, we can not add this method as private method to any specific class.

This method is part of module scope and we are not exporting this method.

File: src/transaction.ts [hunk 14]
```diff
@@ -2493,6 +2467,127 @@ promisifyAll(Transaction, {
...
+  });
+
+  const mutation: spannerClient.spanner.v1.IMutation = {
+    [method]: {table, columns, values},
+  };
+  return mutation as spannerClient.spanner.v1.Mutation;
+}
+
+/**
+ * Formats a delete mutation.
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply src/transaction.ts]: Done

File: src/transaction.ts [hunk 15]
```diff
@@ -2493,6 +2467,127 @@ promisifyAll(Transaction, {
   exclude: ['deleteRows', 'insert', 'replace', 'update', 'upsert'],
 });
 
+/**
+ * Formats the mutations.
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply src/transaction.ts]: Done

File: src/database.ts [hunk 16]
```diff
@@ -3210,6 +3212,110 @@ class Database extends common.GrpcServiceObject {
       }
     }
   }
+
+  /**
+   * Get a list of {@link BatchWriteResponse} as a readable object stream.
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply src/database.ts]: Done

File: src/database.ts [hunk 17]
```diff
@@ -3210,6 +3212,110 @@ class Database extends common.GrpcServiceObject {
...
+   * // If you anticipate many results, you can end a stream early to prevent
+   * // unnecessary processing and API requests.
+   * //-
+   * database.batchWrite()
+   *   .on('data', response => {
+   *     this.end();
+   *   });
+   * ```
+   */
+  batchWrite(
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply src/database.ts]: Done

File: samples/batch-write.js [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+            response.commitTimestamp
+          ).toJSON()}`
+        );
+      })
+      .on('end', () => {
+        console.log('Mutations applied successfully');
+      });
+  } catch (err) {
+    console.log(
+      `Mutation group indexes ${err.indexes} could not be applied with error code ${err.status.code} and error message ${err.status.message}`
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply samples/batch-write.js]: Updated the logs

File: src/database.ts [hunk 19]
```diff
@@ -3210,6 +3212,110 @@ class Database extends common.GrpcServiceObject {
...
+   *  LastName: 'Richards',
+   *  });
+   *
+   * database.batchWrite([mutationGroup])
+   *   .on('error', console.error)
+   *   .on('data', response => {
+   *        console.log('response: ', response);
+   *   })
+   *   .on('end', () => {
+   *        console.log('Request completed successfully');
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply src/database.ts:3261]: End event is not called when there is only an error event

File: samples/batch-write.js [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+  // Creates a client
+  const spanner = new Spanner({
+    projectId: projectId,
+  });
+
+  // Gets a reference to a Cloud Spanner instance and database
+  const instance = spanner.instance(instanceId);
+  const database = instance.database(databaseId);
+
+  // Create Mutation Groups
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply samples/batch-write.js:48]: I have updated the sample

File: samples/batch-write.js [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+        if (response.status.code === 0) {
+          console.log(
+            `Mutation group indexes ${
+              response.indexes
+            }, have been applied with commit timestamp ${Spanner.timestamp(
+              response.commitTimestamp
+            ).toJSON()}`
+          );
+        } else {
+          console.log(
```
Comment (olavloite) [inline samples/batch-write.js:101]:
```suggestion
          // Mutation groups that fail to commit trigger a response with a non-zero status code.
          console.log(
```

File: samples/system-test/spanner.test.js [hunk 22]
```diff
@@ -967,6 +968,37 @@ describe('Autogenerated Admin Clients', () => {
...
+    // console.log(output);
+    // assert.include(
+    //   output,
+    //   /Mutation group indexes [\d,]+ have been applied with commit timestamp \d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d+Z/
+    // );
+    // assert.include(
+    //   output,
+    //   /Mutation group indexes [\d,]+, could not be applied with error code \d+, and error message .+/
+    // );
+    // assert.include(output, 'Request completed successfully');
```
Comment (olavloite) [inline samples/system-test/spanner.test.js:985]: Remove?

```suggestion
```

File: src/transaction.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (olavloite) [reply src/transaction.ts:2479]: Ack

File: src/transaction.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (olavloite) [reply src/transaction.ts:2517]: Ack

**Review by olavloite — APPROVED** [review]

File: samples/system-test/spanner.test.js (hunk 22 above)
Comment (alkatrivedi) [reply samples/system-test/spanner.test.js:985]: Done

File: samples/batch-write.js [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+  const options = {
+    transactionTag: 'batch-write-tag',
+  };
+
+  try {
+    database
+      .batchWriteAtLeastOnce([mutationGroup1, mutationGroup2], options)
+      .on('error', console.error)
+      .on('data', response => {
+        if (response.status.code === 0) {
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply samples/batch-write.js:92]: Done

File: samples/batch-write.js [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+  // Gets a reference to a Cloud Spanner instance and database
+  const instance = spanner.instance(instanceId);
+  const database = instance.database(databaseId);
+
+  // Create Mutation Groups
+  /**
+   * Mutation Groups - A group of mutations to be committed together.
+   * A group must contain related mutations.
+   * Please see the {@link https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#google.spanner.v1.BatchWriteRequest.MutationGroup}
+   * for more details and example.
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply samples/batch-write.js:53]: Done

File: system-test/spanner.ts (hunk 10 above)
Comment (alkatrivedi) [reply system-test/spanner.ts:4880]: I have added an end-to-end test for the error handling of `batchWrite`.
1. I have added a test for the case when there will be invalid mutation groups in the request. Whenever an invalid mutation group will get pass in the request, it will give the response in stream as 'data' event with an `error code of failure`.
2. We are able to get the error event in the case when we are passing an invalid list of mutation groups.
For example - An empty list of Mutation Group.
